src/borrow_tracker/tree_borrows/tree/tests.rs

```diff
@@ -0,0 +1,674 @@
+    /// This makes `LocStateProtPair` more meaningful than a simple
+    /// `(LocStateProt, LocStateProt)` where the two states are not guaranteed
+    /// to be updated at the same time.
+    /// Some `LocStateProtPair` may be unreachable through normal means
+    /// such as `x: Active, y: Active` in the case of mutually foreign pointers.
+    struct LocStateProtPair {
+        x: LocStateProt,
+        y: LocStateProt,
+    }
+
+    impl LocStateProtPair {
+        fn perform_test_access(&self, acc: &TestAccess, xy_rel: RelPosXY) -> Result<Self, ()> {
+            let (xrel, yrel) = acc.ptr.rel_pair(xy_rel);
+            let x = self.x.perform_access(acc.kind, xrel)?;
+            let y = self.y.perform_access(acc.kind, yrel)?;
+            Ok(Self { x, y })
+        }
+
+        fn end_protector_x(&self, _xy_rel: RelPosXY) -> Result<Self, ()> {
+            let x = self.x.end_protector();
+            let y = self.y.clone();
+            Ok(Self { x, y })
+        }
+
+        fn end_protector_y(&self, _xy_rel: RelPosXY) -> Result<Self, ()> {
+            let x = self.x.clone();
+            let y = self.y.end_protector();
+            Ok(Self { x, y })
+        }
+
+        fn perform_test_event<RetX, RetY>(
+            &self,
+            evt: &TestEvent<RetX, RetY>,
+            xy_rel: RelPosXY,
+        ) -> Result<Self, ()> {
+            match evt {
+                TestEvent::Access(acc) => self.perform_test_access(acc, xy_rel),
+                // The fields of the `Ret` variants just serve to make them
+                // impossible to instanciate via the `!` type; we can
+                // always ignore their value.
+                TestEvent::RetX(_) => self.end_protector_x(xy_rel),
+                TestEvent::RetY(_) => self.end_protector_y(xy_rel),
+            }
+        }
+    }
+
+    impl Exhaustive for LocStateProtPair {
+        fn exhaustive() -> Box<dyn Iterator<Item = Self>> {
+            Box::new(<[LocStateProt; 2]>::exhaustive().map(|[x, y]| Self { x, y }))
+        }
+    }
+
+    impl fmt::Display for LocStateProtPair {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            write!(f, "x:{}, y:{}", self.x, self.y)
+        }
+    }
+
+    /// Arbitrary sequence of accesses, as experienced by two mutually foreign pointers
+    /// to the same location.
+    #[derive(Debug, Clone)]
+    struct OpaqueCode<RetX, RetY> {
+        events: Vec<TestEvent<RetX, RetY>>,
+    }
+
+    impl<RetX, RetY> fmt::Display for OpaqueCode<RetX, RetY> {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            for evt in &self.events {
+                write!(f, "{evt}; ")?;
+            }
+            Ok(())
+        }
+    }
+
+    impl LocStateProtPair {
+        /// List all sequences of operations that lead to non-UB configurations,
+        /// without duplicates (not guaranteed to be the shortest sequence of events).
+        fn applicable_opaque_code_without_ub<RetX, RetY>(
```

Comment:
Is this replacing all that previous stuff about only having a maximum of 4 events or so? I was wondering where that had gone.

Not sure if I understand this fn correctly, is this accurate?
```suggestion
    impl LocStateProtPair {
        /// List all sequences of operations start at `self` and do not cause UB
        /// without duplicates (not guaranteed to be the shortest sequence of events).
        /// Yields the states it reaches, and the sequence of operations that got us there.
        fn applicable_opaque_code_without_ub<RetX, RetY>(
```

Maybe rename it so `all_states_reachable_via_opaque_code` or so?

Reply:
Yes, exactly.

This replaces the previous enumeration of sequences of 4 events because computing the reachable set from one `LocStateProtPair` is (unsurprisingly) much more efficient than returning a set of `OpaqueCode` that is known to reach all states from all starting configurations.